### PR TITLE
[SPARK-38098][PYTHON] Add support for ArrayType of nested StructType to arrow-based conversion

### PR DESCRIPTION
**Pull request by LucaCanali**
### What changes were proposed in this pull request?
This proposes to add support for ArrayType of nested StructType to arrow-based conversion. 
This allows Pandas UDFs, mapInArrow UDFs, and toPandas to operate on columns of type Array of Struct, via arrow serialization.
It appears to me that pyarrow 2.0.0 allows to serialize array of struct (while pyarrow 1.0.0 throws an error for this type of data).

### Why are the changes needed?
This extends the usability of pandas_udf.

### Does this PR introduce _any_ user-facing change?
Pandas_udf and mapInArrow will be able to operate on data of type Array of Struct.
toPandas will be able to use arrow serialization for Array of Struct (when spark.sql.execution.arrow.pyspark.enabled=true)

### How was this patch tested?
A test has been added.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
cc @HyukjinKwon

File: python/pyspark/sql/tests/test_pandas_udf_scalar.py [hunk 1]
```diff
@@ -134,6 +134,24 @@ def test_pandas_udf_nested_arrays(self):
         result = df.select(tokenize("vals").alias("hi"))
         self.assertEqual([Row(hi=[["hi", "boo"]]), Row(hi=[["bye", "boo"]])], result.collect())
 
+    def test_pandas_array_struct(self):
+        @pandas_udf("Array<struct<col1:string, col2:long, col3:double>>")
+        def return_cols(cols):
+            return cols
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_pandas_udf_scalar.py:140]: Can we add an assert on its data type here too?

File: python/pyspark/sql/tests/test_pandas_udf_scalar.py [hunk 2]
```diff
@@ -134,6 +134,24 @@ def test_pandas_udf_nested_arrays(self):
...
+            [[[("a", 2, 3.0), ("a", 2, 3.0)]], [[("b", 5, 6.0), ("b", 5, 6.0)]]],
+            "array_struct_col Array<struct<col1:string, col2:long, col3:double>>",
+        )
+        result = df.select(return_cols("array_struct_col").alias("output"))
+        self.assertEqual(
+            [
+                Row(output=[Row(col1="a", col2=2, col3=3.0), Row(col1="a", col2=2, col3=3.0)]),
+                Row(output=[Row(col1="b", col2=5, col3=6.0), Row(col1="b", col2=5, col3=6.0)]),
+            ],
+            result.collect(),
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_pandas_udf_scalar.py:152]: Maybe we will have to also test `toPandas`

File: python/pyspark/sql/tests/test_pandas_udf_scalar.py (hunk 1 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/tests/test_pandas_udf_scalar.py:140]: You can refer to https://github.com/apache/spark/pull/33876 about where to add the tests.

File: python/pyspark/sql/tests/test_pandas_udf_scalar.py (hunk 2 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/tests/test_pandas_udf_scalar.py:152]: You can refer to https://github.com/apache/spark/pull/33876 about where to add the tests.

**Review by HyukjinKwon — COMMENTED** [review]
We would also have to fix documentation:
- https://github.com/apache/spark/blob/master/python/docs/source/user_guide/sql/arrow_pandas.rst#supported-sql-types
- https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L2359-L2364

**Comment by HyukjinKwon** [comment]
cc @BryanCutler @ueshin FYI

File: python/pyspark/sql/pandas/types.py [hunk 3]
```diff
@@ -86,8 +86,13 @@ def to_arrow_type(dt: DataType) -> "pa.DataType":
     elif type(dt) == DayTimeIntervalType:
         arrow_type = pa.duration("us")
     elif type(dt) == ArrayType:
-        if type(dt.elementType) in [StructType, TimestampType]:
+        if type(dt.elementType) == TimestampType:
```
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/types.py:89]: My only concern here is that we current convert the struct type as a DataFrame instead a Series of dictionaries (e.g., `returnType=struct<...>`) which is inconsistent with the current conversion this PR proposes (a Series of dictionaries).

File: python/pyspark/sql/tests/test_pandas_udf_scalar.py [hunk 4]
```diff
@@ -134,6 +134,30 @@ def test_pandas_udf_nested_arrays(self):
...
 
+    def test_pandas_array_struct(self):
+        # SPARK-38098: Support Array of Struct for Pandas UDFs and toPandas
+        import numpy as np
+
+        @pandas_udf("Array<struct<col1:string, col2:long, col3:double>>")
+        def return_cols(cols):
+            self.assertEqual(type(cols), pd.Series)
+            self.assertEqual(type(cols[0]), np.ndarray)
+            self.assertEqual(type(cols[0][0]), dict)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_pandas_udf_scalar.py:145]: I wonder if it makes more sense if we can return a NumPy array of DataFrame here.

File: python/pyspark/sql/pandas/types.py (hunk 3 above)
Comment (LucaCanali) [reply python/pyspark/sql/pandas/types.py:89]: Array of StructType now follows the same type of serialization as StructType.

Comment (LucaCanali) [reply python/pyspark/sql/pandas/types.py:89]: Array of StructType now follows the same type of serialization as StructType.

**Comment by HyukjinKwon** [comment]
@LucaCanali thanks for addressing my comment. The test seems like failing:

```

======================================================================
ERROR [0.182s]: test_pandas_array_struct (pyspark.sql.tests.test_pandas_udf_scalar.ScalarPandasUDFTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/__w/spark/spark/python/pyspark/sql/tests/test_pandas_udf_scalar.py", line 152, in test_pandas_array_struct
    result = df.select(return_cols("array_struct_col"))
  File "/__w/spark/spark/python/pyspark/sql/udf.py", line 276, in wrapper
    return self(*args)
  File "/__w/spark/spark/python/pyspark/sql/udf.py", line 249, in __call__
    judf = self._judf
  File "/__w/spark/spark/python/pyspark/sql/udf.py", line 215, in _judf
    self._judf_placeholder = self._create_judf(self.func)
  File "/__w/spark/spark/python/pyspark/sql/udf.py", line 224, in _create_judf
    wrapped_func = _wrap_function(sc, func, self.returnType)
  File "/__w/spark/spark/python/pyspark/sql/udf.py", line 50, in _wrap_function
    pickled_command, broadcast_vars, env, includes = _prepare_for_python_RDD(sc, command)
  File "/__w/spark/spark/python/pyspark/rdd.py", line 3350, in _prepare_for_python_RDD
    pickled_command = ser.dumps(command)
  File "/__w/spark/spark/python/pyspark/serializers.py", line 458, in dumps
    return cloudpickle.dumps(obj, pickle_protocol)
  File "/__w/spark/spark/python/pyspark/cloudpickle/cloudpickle_fast.py", line 73, in dumps
    cp.dump(obj)
  File "/__w/spark/spark/python/pyspark/cloudpickle/cloudpickle_fast.py", line 602, in dump
    return Pickler.dump(self, obj)
  File "/__w/spark/spark/python/pyspark/cloudpickle/cloudpickle_fast.py", line 316, in _file_reduce
    raise pickle.PicklingError(
_pickle.PicklingError: Cannot pickle files that are not opened for reading: w
```

Could we fix this one?

**Comment by LucaCanali** [comment]
I must say I am bit puzzled by the error found in test_pandas_array_struct as I cannot reproduce it in my test system.
When I run `python/run-tests --modules pyspark-sql --testnames pyspark.sql.tests.test_pandas_udf_scalar` locally I can see the test going through OK.

**Comment by LucaCanali** [comment]
This should be good to go now, @HyukjinKwon ?

File: python/pyspark/sql/tests/test_dataframe.py [hunk 5]
```diff
@@ -953,6 +953,30 @@ def test_to_pandas_from_mixed_dataframe(self):
...
+        # SPARK-38098: Support Array of Struct for Pandas UDFs and toPandas
+        import numpy as np
+        import pandas as pd
+
+        df = self.spark.createDataFrame(
+            [[[("a", 2, 3.0), ("a", 2, 3.0)]], [[("b", 5, 6.0), ("b", 5, 6.0)]]],
+            "array_struct_col Array<struct<col1:string, col2:long, col3:double>>",
+        )
+        is_arrow_enabled = [True, False]
+        for value in is_arrow_enabled:
```
Comment (ueshin) [inline python/pyspark/sql/tests/test_dataframe.py:970]: nit:

```py
for is_arrow_enabled in [True, False]:
```

File: python/pyspark/sql/tests/test_pandas_udf_scalar.py [hunk 6]
```diff
@@ -134,6 +134,30 @@ def test_pandas_udf_nested_arrays(self):
...
 
+    def test_pandas_array_struct(self):
+        # SPARK-38098: Support Array of Struct for Pandas UDFs and toPandas
+        # import numpy as np
+
+        @pandas_udf("Array<struct<col1:string, col2:long, col3:double>>")
+        def return_cols(cols):
+            # self.assertEqual(type(cols), pd.Series)
+            # self.assertEqual(type(cols[0]), np.ndarray)
+            # self.assertEqual(type(cols[0][0]), dict)
```
Comment (ueshin) [inline python/pyspark/sql/tests/test_pandas_udf_scalar.py:145]: I guess we can't use `self` in the udf. Shall we follow the other tests to use builtin `assert` instead:

https://github.com/apache/spark/blob/f8b3d5322e6cbce2e42a6940518686b7255e79cb/python/pyspark/sql/tests/test_pandas_udf_scalar.py#L1206-L1215

Also `import numpy as np` might need to be in the udf.

Comment (LucaCanali) [reply python/pyspark/sql/tests/test_pandas_udf_scalar.py]: Thank you @ueshin for the review and comments.
I have added the proposed modifications.
As for `import numpy as np`, I have now added it explicitly, however externally to the udf, just to be consistent with the other tests there that use numpy.

**Comment by LucaCanali** [comment]
Are there any further comments on this PR?

File: python/pyspark/sql/pandas/types.py [hunk 7]
```diff
@@ -86,9 +86,23 @@ def to_arrow_type(dt: DataType) -> "pa.DataType":
...
+                    "Array of StructType is only supported with pyarrow 2.0.0 and above"
+                )
+            dt_nested = dt.elementType
+            if any(type(field.dataType) == StructType for field in dt_nested):
+                raise TypeError("Nested StructType not supported in conversion to Arrow")
+            fields = [
+                pa.field(field.name, to_arrow_type(field.dataType), nullable=field.nullable)
+                for field in dt_nested
+            ]
+            arrow_type = pa.list_(pa.struct(fields))
```
Comment (ueshin) [inline python/pyspark/sql/pandas/types.py:103]: We can just use `arrow_type = pa.list_(to_arrow_type(dt.elementType))` here?

**Review by ueshin — APPROVED** [review]
Otherwise, LGTM.
@HyukjinKwon Could you take another look? Thanks.

File: python/pyspark/sql/tests/test_dataframe.py [hunk 8]
```diff
@@ -1107,6 +1107,29 @@ def test_to_pandas_from_mixed_dataframe(self):
                 pdf_with_only_nulls = self.spark.sql(sql).filter("tinyint is null").toPandas()
                 self.assertTrue(np.all(pdf_with_only_nulls.dtypes == pdf_with_some_nulls.dtypes))
 
+    @unittest.skipIf(
+        not have_pandas or not have_pyarrow,
+        cast(str, pandas_requirement_message or pyarrow_requirement_message),
+    )
+    def test_to_pandas_for_array_of_struct(self):
```
Comment (ueshin) [inline python/pyspark/sql/tests/test_dataframe.py:1114]: Could you mark to skip this test if the underlying PyArrow is less than `2.0.0`?

File: python/pyspark/sql/tests/test_pandas_udf_scalar.py [hunk 9]
```diff
@@ -134,6 +134,30 @@ def test_pandas_udf_nested_arrays(self):
         result = df.select(tokenize("vals").alias("hi"))
         self.assertEqual([Row(hi=[["hi", "boo"]]), Row(hi=[["bye", "boo"]])], result.collect())
 
+    def test_pandas_array_struct(self):
```
Comment (ueshin) [inline python/pyspark/sql/tests/test_pandas_udf_scalar.py:137]: ditto.

**Review by HyukjinKwon — COMMENTED** [review]
I am fine w/ this. cc @BryanCutler too FYI

**Review by BryanCutler — APPROVED** [review]
LGTM, thanks @LucaCanali !

**Comment by HyukjinKwon** [comment]
Alright, let's merge this in once the tests pass. @LucaCanali mind fixing the lint failure?

```

starting mypy annotations test...
annotations failed mypy checks:
python/pyspark/sql/tests/test_dataframe.py:1113: error: Redundant cast to "str"  [redundant-cast]
Found 1 error in 1 file (checked 341 source files)
1
```

**Comment by LucaCanali** [comment]
Thank you @HyukjinKwon @ueshin and @BryanCutler

**Comment by BryanCutler** [comment]
Tests passing now, merged to master. Thanks all!

File: python/pyspark/sql/pandas/types.py [hunk 10]
```diff
@@ -86,8 +86,15 @@ def to_arrow_type(dt: DataType) -> "pa.DataType":
     elif type(dt) == DayTimeIntervalType:
         arrow_type = pa.duration("us")
     elif type(dt) == ArrayType:
-        if type(dt.elementType) in [StructType, TimestampType]:
+        if type(dt.elementType) == TimestampType:
             raise TypeError("Unsupported type in conversion to Arrow: " + str(dt))
+        elif type(dt.elementType) == StructType:
+            if LooseVersion(pa.__version__) < LooseVersion("2.0.0"):
+                raise TypeError(
+                    "Array of StructType is only supported with pyarrow 2.0.0 and above"
```
Comment (itholic) [inline python/pyspark/sql/pandas/types.py:94]: Hi, @LucaCanali I think I've found a case where Array of StructType doesn't work properly:

**In:**
```python
df = spark.createDataFrame(
  [
    ("a", [("b", False), ("c", True)]),
  ]
).toDF("c1", "c2")
df.toPandas()
```

**Out:**
```python
  c1                                                 c2
0  a  [{'_1': 'b', '_2': False}, {'_1': 'c', '_2': T...
```

**Expected:**
```python
  c1                       c2
0  a  [(b, False), (c, True)]
```


I roughly suspect that this is maybe an internal table conversion issue from PyArrow side but not very sure about this, so could you check this out when you find some time??

Thanks in advance!

Comment (LucaCanali) [reply python/pyspark/sql/pandas/types.py:94]: Hi, @itholic thank you for reporting this.
I have tried to reproduce the reported issue in my environment, but I get the expected result instead (I use Python 3.11, Spark 3.5.1, pyarrow 11). Can you please confirm that you still have this issue?

**Pull request closed without merging**